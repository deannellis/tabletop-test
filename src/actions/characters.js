import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

