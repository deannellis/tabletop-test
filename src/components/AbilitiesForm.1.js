import React from 'react';

export default class AbilitiesForm extends React.Component {
    state = {
        strength: undefined,
        intelligence: undefined,
        wisdom: undefined,
        dexterity: undefined,
        constitution: undefined,
        charisma: undefined
    };
    
    renderButtons = () => {
        Object.keys(this.state).map((key) => {
            return 
            <div>
                <button type="button" onClick={this.onRollStrength} disabled={typeof this.state.strength != 'undefined'}>Strength</button>
                <p>Strength: {typeof this.state.strength == 'undefined' ? 'Roll to determine Strength' : this.state.strength }</p>
            
        });
    };
    
    onRollStrength = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({strength: rollValue}));
    };
    
    onRollIntelligence = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({intelligence: rollValue}));
    };
    
    onRollWisdom = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({wisdom: rollValue}));
    };
    
    onRollDexterity = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({dexterity: rollValue}));
    };
    
    onRollConstitution = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({constitution: rollValue}));
    };
    
    render() {
        return (
            <div>
                <form>
                    {this.renderButtons()}
                </form>
            </div>
        );
    }
}