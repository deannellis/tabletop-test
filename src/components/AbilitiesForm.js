import React from 'react';
import SingleRollDie from './SingleRollDie';
import uuid from 'uuid';

export default class AbilitiesForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            strength: [],
            intelligence: [],
            wisdom: [],
            dexterity: [],
            constitution: [],
            charisma: [],
            error: '',
            refreshKey: 0,
            lastDieRolled: undefined,
            lastResult: undefined
        };
    }
    
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {this.renderAttribute('Strength', 'strength')}
                    {this.renderAttribute('Intelligence', 'intelligence')}
                    {this.renderAttribute('Wisdom', 'wisdom')}
                    {this.renderAttribute('Dexterity', 'dexterity')}
                    {this.renderAttribute('Constitution', 'constitution')}
                    {this.renderAttribute('Charisma', 'charisma')}
                    <button>Next Step</button>
                </form>
            </div>
        );
    }
    
    handleRollDie = (result, attributeKey) => {
        this.setState(() => ({
            lastResult: result,
            [attributeKey]: [...this.state[attributeKey], result]
        }));
    };
    
    onSubmit = (e) => {
        e.preventDefault();
        
        const sum = (accumulator, currentValue) => accumulator + currentValue;
        
        if(this.state.strength.length < 3 || this.state.intelligence.length < 3 || this.state.wisdom.length < 3 || this.state.dexterity.length < 3 || this.state.constitution.length < 3 || this.state.charisma.length < 3) {
            this.setState(() => ({
                error: 'Please roll all scores'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                id: uuid(),
                abilities: {
                    strength: this.state.strength.reduce(sum),
                    intelligence: this.state.intelligence.reduce(sum),
                    wisdom: this.state.wisdom.reduce(sum),
                    dexterity: this.state.dexterity.reduce(sum),
                    constitution: this.state.constitution.reduce(sum),
                    charisma: this.state.charisma.reduce(sum)
                },
                name: 'In-Progress Character',
                charClass: 'unkown'
            });
        } 
    }
    
    renderAttribute = (label, attributeKey) => {
        const attributeValue = this.state[attributeKey];
        const isDefined = attributeValue.length !== 0;
        let rollsString = '';
        if(this.state[attributeKey].length < 3){
            this.state[attributeKey].forEach(
                (roll) => {rollsString += `${roll} + `}
            );
        } else {
            let total = 0;
            this.state[attributeKey].forEach(
                (roll) => {total += roll}
            );
            rollsString = `${this.state[attributeKey][0]} + ${this.state[attributeKey][1]} + ${this.state[attributeKey][2]} = ${total}`;
        }
        
        const message = label + ': ' + (isDefined ? rollsString : 'roll to determine');
        return(
            <div>
                <p>{message}</p>
                <div className="dice-container">
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                </div>
            </div>
        );
    }
    
    updateRefreshKey = (newKey) => {
        this.setState(() => ({
            refreshKey: newKey
        }));
    };
}