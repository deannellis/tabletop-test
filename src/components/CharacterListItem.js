import React from 'react';
import { Link } from 'react-router-dom';

const CharacterListItem = ({ name, charClass, race, id }) => (
    <div>
        <h2>{name}</h2>
        <p>{race} - {charClass}</p>
        <Link to={`/view-char/${id}`}>View Character</Link>
    </div>    
);

export default CharacterListItem;
