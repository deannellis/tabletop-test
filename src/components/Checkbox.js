import React, { Component, PropTypes } from 'react';

class Checkbox extends React.Component {
    state = {
        isChecked: false,
    }
    
    toggleCheckboxChange = () => {
        const { handleCheckboxChange } = this.props;
        const { isChecked } = this.state;
    
        this.setState(({ isChecked }) => (
            {
                isChecked: !isChecked,
            }
        ));
    
        handleCheckboxChange(!isChecked);
    }
    
      render() {
        const { label } = this.props;
        const { isChecked } = this.state;
        let isDisabled;
        if(isChecked) {
            isDisabled = false;
        } else {
            isDisabled = this.props.isDisabled
        }
    
        return (
          <div className="checkbox">
            <label>
                <input
                    type="checkbox"
                    value={isChecked}
                    checked={isChecked}
                    onChange={this.toggleCheckboxChange}
                    disabled={isDisabled}
                />
    
              {label}
            </label>
          </div>
        );
      }
}

export default Checkbox;