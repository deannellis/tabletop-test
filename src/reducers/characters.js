//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};