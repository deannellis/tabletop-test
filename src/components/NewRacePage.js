import React from 'react';
import { connect } from 'react-redux';
import RaceForm from './RaceForm';
import { editCharacter } from '../actions/characters';

const NewRacePage = (props) => (
    <div>
        <RaceForm 
            onSubmit={(id, character) => {
                props.dispatch(editCharacter(id, character));
                props.history.push('/new-char-step-3');
            }}
        />
    </div>
);

export default connect()(NewRacePage);