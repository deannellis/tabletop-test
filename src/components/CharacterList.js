import React from 'react';
import { connect } from 'react-redux';
import CharacterListItem from './CharacterListItem';

const CharacterList = (props) => (
    <div>
        <h1>Character List</h1>
        {props.characters.length > 0 ?
            <div>
                {props.characters.map((character) => {
                    return <CharacterListItem key={character.id} {...character} />
                })}
            </div>
        :
            <p>You have no characters, Click below to create a character</p>
        }
    </div>    
);

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(CharacterList);
