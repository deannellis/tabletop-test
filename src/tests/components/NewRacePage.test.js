/* global jest expect */
import React from 'react';
import { shallow } from 'enzyme';
import { NewRacePage } from '../../components/NewRacePage';
import characters from '../fixtures/characters';

test('should render NewRacePage correctly', () => {
    const onSubmit = jest.fn();
    const history = { push: jest.fn() }
    const wrapper = shallow(<NewRacePage onSubmit={onSubmit} history={history} />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
    const onSubmit = jest.fn();
    const history = { push: jest.fn() }
    const wrapper = shallow(<NewRacePage onSubmit={onSubmit} history={history} />);
    wrapper.find('RaceForm').prop('onSubmit')(characters[1].id, characters[1]);
    expect(history.push).toHaveBeenLastCalledWith(`/new-char-step-3/${characters[1].id}`);
    expect(onSubmit).toHaveBeenLastCalledWith(characters[1].id, characters[1]);
});