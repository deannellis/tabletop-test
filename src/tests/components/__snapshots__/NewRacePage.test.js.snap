// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render NewRacePage correctly 1`] = `
<div>
  <RaceForm
    onSubmit={[Function]}
  />
</div>
`;
